lib/vtls/openssl.c

```diff
@@ -258,6 +263,20 @@
 #define HAVE_OPENSSL_VERSION
 #endif
 
+struct ossl_x509_store {
+#if defined(HAVE_THREADS)
+  curl_mutex_t mtx;  /* protects access to `CAfile` and `store` */
+#endif
+  char *CAfile;      /* CAfile path used to generate X509 store */
+  X509_STORE *store; /* cached X509 store or NULL if none */
+};
+
+struct ossl_global_data {
+  struct ossl_x509_store x509_ctx; /* X509 store caching context */
+};
+
+static struct ossl_global_data ossl_global;
```

Comment:
I am not fond of using globals for this functionality.

Reply:
Yeah, I was looking for alternatives for where to store the cached `X509_STORE`, but didn't see anything obvious.

I can get to the `struct Curl_multi` for the connection which would let me store it there, fixing it only for applications that use the multi interface. (NetSurf does this, but I wanted something more general.)

I'm not sure what the `struct Curl_share` is though. Maybe I should look more into that?

I'm not sure how we can store a shared `X509_STORE` without *some* form of mutex protections though.

Do you have any suggestions?

Reply:
It's now stored on the multi handle.